streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -190,11 +190,42 @@ public void print(Serde<K> keySerde, Serde<V> valSerde) {
 
     @Override
     public void print(Serde<K> keySerde, Serde<V> valSerde, String streamName) {
+        print(keySerde, valSerde, streamName, null);
+    }
+
+    @Override
+    public void print(final Serde<K> keySerde, final Serde<V> valSerde, String streamName, final KeyValueMapper<K, V, String> mapper){
         String name = topology.newName(PRINTING_NAME);
         streamName = (streamName == null) ? this.name : streamName;
-        topology.addProcessor(name, new KeyValuePrinter<>(keySerde, valSerde, streamName), this.name);
+        topology.addProcessor(name, new KStreamPeek<>(printAction(System.out, keySerde, valSerde, streamName, mapper)), this.name);
     }
 
+    private static <K, V> ForeachAction<K, V> printAction(final PrintStream printStream, final Serde<?> keySerde, final Serde<?> valueSerde, final String streamName, final KeyValueMapper<K, V, String> mapper) {
+        return new ForeachAction<K, V>() {
+            @Override
+            public void apply(final K key, final V value) {
+                K keyToPrint = (K) maybeDeserialize(key, keySerde.deserializer());
+                V valueToPrint = (V) maybeDeserialize(value, valueSerde.deserializer());
+                if(mapper == null) {
+                    printStream.println("[" + streamName + "]: " + keyToPrint + " , " + valueToPrint);
+                } else {
+                    printStream.println("[" + streamName + "]: " + mapper.apply(keyToPrint, valueToPrint));
+                }
+            }
+
+            private Object maybeDeserialize(Object receivedElement, Deserializer<?> deserializer) {
+                if (receivedElement == null) {
+                    return null;
+                }
+
+                if (receivedElement instanceof byte[]) {
+                    return deserializer.deserialize("Topic", (byte[]) receivedElement);
```

Comment:
@guozhangwang maybe I'm missing something, but is there any way to infer the `topic` at this stage?
Otherwise, indeed, this needs to be moved to a extension of `KStreamPeek` in order to have access to `context.topic()`.
Couldn't there be more use cases where one would appreciate to have access the context where this will be processed?

I think after this is clear to me, I'll be able to reopen the separate PR.